redirector.rs/src/main.rs

```diff
@@ -0,0 +1,180 @@
+            let target = mountpoint.join(subdir);
+            reply.data(target.as_os_str().as_bytes());
+        } else {
+            reply.error(ENOENT);
+        }
+    }
+
+    fn readdir(
+        &mut self,
+        _req: &Request,
+        ino: u64,
+        _fh: u64,
+        offset: i64,
+        mut reply: ReplyDirectory,
+    ) {
+        if ino == ROOT_DIR_INODE {
+            if offset == 0 {
+                reply.add(ROOT_DIR_INODE, 0, FileType::Directory, ".");
+                reply.add(ROOT_DIR_INODE, 1, FileType::Directory, "..");
+                reply.add(PRIVATE_LINK_INODE, 2, FileType::Symlink, "private");
+                reply.add(PUBLIC_LINK_INODE, 3, FileType::Symlink, "public");
+                reply.add(TEAM_LINK_INODE, 4, FileType::Symlink, "team");
+            }
+            reply.ok();
+        } else {
+            reply.error(ENOENT);
+        }
+    }
+}
+
+// This is explicitly only implemented for Linux, so that building on other
+// platforms is a compiler error.
+#[cfg(target_os = "linux")]
+fn get_keybase_fuse_mount(uid: u32) -> Option<PathBuf> {
+    let uid_str = format!("user_id={}", uid);
+    let file = File::open("/proc/mounts").expect("failed to open /proc/mounts");
+    let reader = BufReader::new(file);
+    // TODO: Avoid assuming UTF8?
+    // TODO: Sort the paths first?
+    for line in reader.lines() {
+        let line = line.expect("failed to read from /proc/mounts");
+        let parts: Vec<&str> = line.split_whitespace().collect();
+        // Check the mount type.
+        if parts[2] != "fuse" {
+            continue;
+        }
+        // Look for the uid in the mount options.
+        if parts[3].split(',').find(|s| s == &uid_str).is_none() {
+            continue;
+        }
+        // Make sure "keybase" is in the path somewhere. We do this loosely,
+        // rather than splitting exactly on slashes, to handle "keybase.devel"
+        // etc.
+        if parts[1].to_lowercase().find("keybase").is_none() {
+            continue;
+        }
+        let path: PathBuf = parts[1].into();
+        // TODO: We want to check that .kbfs_error exists, but the kbfsfuse
+        // mount doesn't currently set "allow_root".
+        return Some(path);
+    }
+    None
+}
+
+fn main() {
+    // The rust-fuse library writes logs using the standard log crate
+    // interface. To init a log backend that will actually print those,
+    // uncomment this line:
+    // env_logger::Builder::new().filter(None, log::LevelFilter::Debug).init();
+
+    let redirector = Redirector {
+        start_time: time::get_time(),
+    };
+    let mountpoint = env::args_os().nth(1).expect("must have a path argument");
+    let options = &["-o", "allow_other", "-o", "fsname=keybase-redirector"];
+    let options_osstr: Vec<&OsStr> = options.iter().map(|o| o.as_ref()).collect();
+    fuse::mount(redirector, &mountpoint, &options_osstr).expect("mount returned an error");
```

Comment:
Also we probably need to either do an flock around here to make sure that only one process gets to do the mount, and the others just silently exit; or we should just ignore (and log?) the specific failure that happens if it's already mounted.  Any local user should be able to call `run_keybase` (which invokes the redirector) and it should just work.

Reply:
If `run_keybase` is invoking the redirector with redirected output and a backgrounding `&`, does it matter what error code the redirector exits with? Maybe we want to make sure the log file doesn't grow in this case?

Reply:
Hmm good point.  I guuess `set -e` scripts wouldn't fail on a non-zero exit code if the call is backgrounded.  Still, you make a good point about the log, and it'd be nice if it would work well even without  the backgrounding (say if we ever got around to systemd-ing it).